README.md
![AgroPhos ERP](./themes/default/images/logo_simple_erp.jpeg  "AgroPhos ERP")
===================

AgroPhos ERP is open source, web-based accounting software for small and medium enterprises.
It supports double entry accounting providing both low level journal entry and user friendly, document based 
interface for everyday business activity with automatic GL postings generation. This is multicurrency,
multilanguage system with active worldwide users community:

* [Project web site](http://AgroPhos.com)
* [SourceForge project page](http://sourceforge.net/projects/AgroPhos/)
* [Central users forum](http://AgroPhos.com/punbb/index.php)
* [Main code repository](https://sourceforge.net/p/AgroPhos/git/ci/master/tree/)
* [GitHub mirror](http://github.com/AgroPhosERP/FA)
* [Mantis bugtracker](http://mantis.AgroPhos.com)
* [AgroPhos Wiki](http://AgroPhos.com/fawiki/)

This project is developed as cooperative effort by AgroPhos team and available under [GPL v.3 license](./doc/license.txt) 

## Requirements

To use AgroPhos application you should have already installed: 

*   Any HTTP web server supporting php eg. _**Apache with mod_php**_ or _**IIS**_.
*   **_PHP_** >=5.0 (version 5.6 or 7.x is recommended)
*   **_MySQL_** >=4.1 server with **_Innodb_** tables enabled, or any version on **MariaDB** server
*   **_Adobe Acrobat Reader_** (or any another PDF reader like _**evince**_) is handy for viewing reports before printing them out.

## Installation
### 1. PHP configuration checks

*   One critical aspect of the PHP installation is the setting of **_session.auto_start_** in the php.ini file. Some rpm distributions of PHP have the default setting of **_session.auto_start = 1_**. This starts a new session at the beginning of each script. However, this makes it impossible to instantiate any class objects that the system relies on. Classes are used extensively by this system. When sessions are required they are started by the system and this setting of **_session.auto_start_** can and should be set to 0.
*   For security reasons both Register Globals and Magic Quotes php settings should be set to Off. When AgroPhos is used with www server running php as Apache module, respective flags are set in .htaccess file. When your server uses CGI interface to PHP you should set  **_magic_quotes_gpc = 0_** and **_register_globals = 0_** in php.ini file.
*   **_Innodb_** tables must be enabled in the MySQL server. These tables allow database transactions which are a critical component of the software. This is enabled by default in the newer versions of MySQL. If you need to enable it yourself, consult the MySQL manual.

### 2. Download application files

* Download and unpack latest AgroPhos tarball from SourceForge into folder created under web server document root, e.g. **/var/www/html/AgroPhos**

* If you prefer easy upgrades when new minor versions are released, you can clone sources from SourceForge project page or Github mirror e.g.:
>	# cd  /var/www/html
>	# git clone `https://git.code.sf.net/p/AgroPhos/git` AgroPhos

Master branch contains all the latest bugfixes made atop the last stable release.
	
### 3. Installation

Use your browser to open page at URL related to chosen installation folder. As an example, if you plan to use application locally and in previous step you have put files on your Linux box in /var/www/html/AgroPhos subfolder, just select `http://localhost/AgroPhos` url in your browser, and you will see start page of installation wizard. Follow instructions displayed during the process.

During installation you will need to provide data server credentials with permissions to create new database, or you will have to provide existing database name and credentials for user with valid usage permissions to access it. You will have to chose also a couple of other options including installation language, optimal encoding for database data etc. Keep in mind that some options (like additional translations and charts of accounts) presented during installation process could be installed also later, when AgroPhos is already in use.

 After successful installation please remove or rename your install directory for safety reasons. You won't need it any more.

### 4. Logging In For the First Time

 Open a browser and enter the URL for the web server directory where AgroPhos is installed. Enter the user name  **admin** and use password declared during install process to login as company administrator. Now you can proceed with configuration process setting up additional user accounts, creating fiscal years, defining additional currencies, GL accounts etc. All configuration options available in application are described in [AgroPhos Wiki](http://AgroPhos.com/fawiki/) available directly from Help links on every application page under ![Help](./themes/default/images/help.gif  "Help") icon.
 

## Troubleshooting

If you encountered any problems with AgroPhos configuration or usage, please consult your case with other users on [AgroPhos forum](http://AgroPhos.com/punbb/index.php). If you think you have encountered a bug in application and after consulting other community members you still are sure this is really a bug, please fill in a report in project [Mantis bugtracker](http://mantis.AgroPhos.com) with all details which allow development team reproduce the problem, and hopefully fix it. Keep in mind, that  [GitHub](http://github.com/AgroPhosERP/FA) page is mainly passive mirror for project based on SorceForge, so posting bug reports here is at least suboptimal.


CHANGELOG.txt
2020-02-04 14:09:15 +0100 Joe Hunt 
Release 2.4.8
M	CHANGELOG.txt
M	version.php

2020-02-04 13:07:00 +0100 Janusz Dobrowolski 
Gettext template updated.
M	lang/new_language_template/LC_MESSAGES/empty.po

2020-02-04 09:36:12 +0100 Joe Hunt 
Updated en_US-demo.sql and en_US-new.sql COAs (moved dates 1 year forwards)
M	sql/en_US-demo.sql
M	sql/en_US-new.sql

2020-01-29 08:49:29 +0100 Janusz Dobrowolski 
[0004904] Customer Credit Note: fixed invalid inventory GL postings for service items.
M	sales/includes/db/sales_credit_db.inc

2020-01-25 11:05:14 +0100 Janusz Dobrowolski 
Backup and Restore: Added support for views (@oakstreet1)
M	admin/db/maintenance_db.inc
M	includes/db/connect_db.inc

2020-01-25 10:52:32 +0100 Janusz Dobrowolski 
Added missing new install translation files.
A	install/lang/nb_NO/LC_MESSAGES/nb_NO.mo
A	install/lang/nb_NO/LC_MESSAGES/nb_NO.po
A	install/lang/nl_NL/LC_MESSAGES/nl_NL.mo
A	install/lang/nl_NL/LC_MESSAGES/nl_NL.po

2020-01-23 22:51:12 +0100 Joe Hunt 
Optimizing the F2, F3 and F4 keys in sales/purchase forms.
M	includes/ui/ui_controls.inc
M	inventory/manage/items.php
M	purchasing/manage/suppliers.php
M	sales/manage/customers.php

2020-01-23 22:05:05 +0100 Janusz Dobrowolski 
Added nb_NO and nl_NL installer translations by Stefan Nijessen.
M	install/isession.inc

2020-01-21 10:13:37 +0100 Joe Hunt 
F2 - Customers/Suppliers; F4 - Items. Removed Menu in popup windows.
M	inventory/manage/items.php
M	purchasing/manage/suppliers.php
M	sales/manage/customers.php

2020-01-18 09:03:18 +0100 Janusz Dobrowolski 
Fixed Assets Depreciation: removed invalid and obsolete audit record update.
M	fixed_assets/includes/fixed_assets_db.inc

2020-01-16 09:10:56 +0100 Joe Hunt 
Uninitialized string offset in main.inc Fixed for php >= 7.0
M	includes/main.inc

2020-01-16 08:57:36 +0100 Joe Hunt 
Tax rate displayed twice in rep209 (Purchase orders). Fixed.
M	reporting/rep209.php

2020-01-13 15:53:39 +0100 Joe Hunt 
Sales Order Inquiry Sort Descenting?, Fixed.
M	sales/includes/db/sales_order_db.inc
M	sales/inquiry/sales_orders_view.php

2020-01-01 16:51:21 +0100 Joe Hunt 
Amendment Required in rep601.php for voided transactions. Fixed.
M	reporting/rep601.php

2019-12-29 09:09:40 +0100 Joe Hunt 
bank statement report with 'All' Selection of Bank Accounts fixed
M	reporting/includes/reports_classes.inc
M	reporting/rep601.php
M	reporting/reports_main.php

2019-12-26 07:50:36 +0100 Joe Hunt 
Customer Payment Entry - Scrolling behaviour when clicking All or None. Fixed.
M	includes/ui/allocation_cart.inc

2019-11-26 13:44:56 +0100 Janusz Dobrowolski 
[0005002] Journal Inquiry: incorrect amounts displayed for journal entries with more than one customer/supplier accounts, fixed.
M	gl/includes/db/gl_db_trans.inc

2019-11-26 07:45:50 +0100 Joe Hunt 
Bug 4988: Auto Increase of Document References cant be disabled. Fixed.
M	sales/includes/cart_class.inc

2019-11-25 14:46:22 +0100 Janusz Dobrowolski 
[0004993] Sales Order: incorrect prepayment guess for wholesale (non tax included) pricelists.
M	sales/includes/cart_class.inc

2019-11-25 08:44:00 +0100 Joe Hunt 
Formatting error in dashboard.inc fixed.
M	includes/dashboard.inc

2019-11-10 01:37:06 +0100 Joe Hunt 
recurring invoices - period improved.
M	sales/create_recurrent_invoices.php

2019-11-10 01:07:53 +0100 Joe Hunt 
count error when printing to Excel. Fixed.
M	reporting/includes/Workbook.php
M	reporting/includes/excel_report.inc
M	reporting/includes/pdf_report.inc

2019-10-28 09:44:33 +0100 Joe Hunt 
Bug 5009: Page # on invoices counts up across invoices. Fixed for all documents.
M	reporting/includes/pdf_report.inc

2019-10-26 08:54:57 +0200 Joe Hunt 
Misspelled parameter in rep115.php
M	reporting/rep115.php

2019-10-21 22:44:07 +0200 Joe Hunt 
Bug 5013: zero invoices create zero payments and cause db error upon view payment. Fixed.
M	sales/includes/db/sales_invoice_db.inc

2019-10-14 01:15:54 +0200 Joe Hunt 
Voided Documents Still Exist In Drop-down when printing. Fixed.
M	reporting/includes/reports_classes.inc

2019-10-11 10:24:50 +0200 Janusz Dobrowolski 
Additional fixes to supplier//customer attachments.
M	admin/attachments.php
M	purchasing/manage/suppliers.php

2019-10-09 23:55:30 +0200 Janusz Dobrowolski 
Added file missing from previous commit.
A	includes/ui/attachment.inc

2019-10-07 23:40:55 +0200 Janusz Dobrowolski 
Added new attachment types for customer and supplier specific documents.
M	admin/attachments.php
M	admin/db/attachments_db.inc
M	admin/db/transactions_db.inc
M	includes/sysnames.inc
M	includes/types.inc
M	purchasing/manage/suppliers.php
M	sales/manage/customers.php

2019-10-06 10:00:28 +0200 Joe Hunt 
reports_classes.inc:40: count() has been fixed.
M	reporting/includes/pdf_report.inc

2019-10-04 09:48:59 +0200 Joe Hunt 
Additional Transaction Date presence check
M	admin/db/transactions_db.inc

2019-09-03 18:04:42 +0200 Joe Hunt 
inventory location transfer from sales kit did not work. Fixed.
M	inventory/includes/stock_transfers_ui.inc

2019-08-28 17:24:52 +0200 Joe Hunt 
Database error message in sales_order_db.inc fixed.
M	sales/includes/db/sales_order_db.inc

2019-08-17 17:39:07 +0200 Joe Hunt 
Clone Manufactured Item -> Clone BOM.
M	manufacturing/manage/bom_edit.php

2019-08-17 17:38:19 +0200 Joe Hunt 
Report Sales Listing fix.
M	reporting/rep106.php

2019-08-17 16:59:41 +0200 Joe Hunt 
Bug 4987: In rtl languages, text aligned centered is printed aligned right in PDF documents. Fixed.
M	reporting/includes/class.pdf.inc

2019-08-08 08:09:28 +0200 Joe Hunt 
Bug 4980: The field characters size of "Memo" in voiding a transaction doesn't correspond to the field size in the database. Fixed by cutting the text to 60 characters in ausit_trail.
M	includes/db/audit_trail_db.inc

2019-08-05 09:29:23 +0200 Joe Hunt 
Rerun - Show more relevant info in CounterPart Field, Journal Inquiry.
M	gl/includes/db/gl_db_trans.inc

2019-08-01 22:57:27 +0200 Joe Hunt 
Show more relevant info in CounterPart Field, Journal Inquiry.

js/inserts.js
/**********************************************************************
    
***********************************************************************/
var _focus;
var _hotkeys = {
	'alt': false,	
	'list': false, 
	'focus': -1		
};

function validate(e) {
	if (e.name && (typeof _validate[e.name] == 'function'))
		return _validate[e.name](e);
	else {
		var n = e.name.indexOf('[');
		if(n!=-1) {
			var key = e.name.substring(n+1, e.name.length-1);
			if (key.length>1 && _validate[e.name.substring(0,n)])
				return _validate[e.name.substring(0,n)][key](e);
		}
	}
	return true;
}

function set_fullmode() {
	document.getElementById('ui_mode').value = 1;
	document.loginform.submit();
	return true;
}

function save_focus(e) {
  _focus = e.name||e.id;
  var h = document.getElementById('hints');
  if (h) {
	h.style.display = e.title && e.title.length ? 'inline' : 'none';
	h.innerHTML = e.title ? e.title : '';
  }
}

function _expand(tabobj) {

  var ul = tabobj.parentNode.parentNode;
  var alltabs=ul.getElementsByTagName("button");
  var frm = tabobj.form;

  if (ul.getAttribute("rel")){
	for (var i=0; i<alltabs.length; i++){
	  alltabs[i].className = "ajaxbutton"  //deselect all tabs 
	}
	tabobj.className = "current";
	JsHttpRequest.request(tabobj)
  }
}

//interface for selecting a tab (plus expand corresponding content)
function expandtab(tabcontentid, tabnumber) {
  var tabs = document.getElementById(tabcontentid);
 _expand(tabs.getElementsByTagName("input")[tabnumber]);
}

function _set_combo_input(e) {
		e.setAttribute('_last', e.value);
		e.onblur=function() {
		  var but_name = this.name.substring(0, this.name.length-4)+'button';
		  var button = document.getElementsByName(but_name)[0];
		  var select = document.getElementsByName(this.getAttribute('rel'))[0];
		  save_focus(select);



		  if (button && (this.value != this.getAttribute('_last'))) {
			JsHttpRequest.request(button);
		  } else if(string_contains(this.className, 'combo2')) {
				this.style.display = 'none';
				select.style.display = 'inline';
				setFocus(select);
		  }
		  return false;
		};
		e.onkeyup = function(ev) {
			var select = document.getElementsByName(this.getAttribute('rel'))[0];
			if(select && select.selectedIndex>=0) {
			  var len = select.length;
			  var byid = string_contains(this.className, 'combo') || string_contains(this.className, 'combo3');
			  var ac = this.value.toUpperCase();
			  select.options[select.selectedIndex].selected = false;
			  for (i = 0; i < len; i++) {
				var txt = byid ? select.options[i].value : select.options[i].text;
				if (string_contains(this.className, 'combo3')) {
				  if(txt.toUpperCase().indexOf(ac) == 0) {
					select.options[i].selected = true;
				  	break;
				  }
				} else {
				  if(txt.toUpperCase().indexOf(ac) >= 0) {
					select.options[i].selected = true;
				  	break;
				  }
				}
			  }
			}
		};
    	e.onkeydown = function(ev) {
	  		ev = ev||window.event;
	  		key = ev.keyCode||ev.which;
	  		if(key == 13) {
			  this.blur();
	  		  return false;
	  		}
		}
}

function _update_box(s) {
	var byid = string_contains(s.className, 'combo') || string_contains(s.className, 'combo3');
	var rel = s.getAttribute('rel');
	var box = document.getElementsByName(rel)[0];
		if(box && s.selectedIndex>=0) {
			  var opt = s.options[s.selectedIndex];
				if(box) {
				  var old = box.value;
				  box.value = byid ? opt.value : opt.text;
				  box.setAttribute('_last', box.value);
				  return old != box.value
				}
		}
}

function _set_combo_select(e) {
		
		
		
		e.setAttribute('_last', e.selectedIndex);
		e.onblur = function() {
		    var box = document.getElementsByName(this.getAttribute('rel'))[0];
			if ((this.selectedIndex != this.getAttribute('_last'))
				||((string_contains(this.className, 'combo') || string_contains(this.className, 'combo3')) && _update_box(this))
				)
					this.onchange();
		}
		e.onchange = function() {
			var s = this;
			this.setAttribute('_last', this.selectedIndex);
			if(string_contains(s.className, 'combo') || string_contains(this.className, 'combo3'))
			    _update_box(s);
			if(s.selectedIndex>=0) {
				 var sname = '_'+s.name+'_update';
				 var update = document.getElementsByName(sname)[0];
				 if(update) {
					    JsHttpRequest.request(update);
				}
			}
			return true;
		}
		e.onkeydown = function(event) {
		    event = event||window.event;
		    key = event.keyCode||event.which;
		    var box = document.getElementsByName(this.getAttribute('rel'))[0];
	  		if(key == 8 || (key==37 && event.altKey)) {
				event.returnValue = false;
  			  	return false;
  			}
		    if (box && (key == 32) && (string_contains(this.className, 'combo2'))) {
			    this.style.display = 'none';
			    box.style.display = 'inline';
				box.value='';
				setFocus(box);
			    return false;
			 } else {
			 	if (key == 13 && !e.length) 
			 		return false;
			 }
		}
}

var _w;

function callEditor(key) {
  var el = document.getElementsByName(editors[key][1])[0];
  if(_w) _w.close(); 
  var left = (screen.width - editors[key][2]) / 2;
  var top = (screen.height - editors[key][3]) / 2;
  _w = open(editors[key][0]+el.value+'&popup=1',
	  "edit","scrollbars=yes,resizable=0,width="+editors[key][2]+",height="+editors[key][3]+",left="+left+",top="+top+",screenX="+left+",screenY="+top);
  if (_w.opener == null)
	  _w.opener = self;
  editors._call = key; 
  _w.focus();
}

function passBack(value) {
	var o = opener;
	if(value != false) {
		var back = o.editors[o.editors._call]; 
		var to = o.document.getElementsByName(back[1])[0];
		if (to) {
			if (to[0] != undefined)
				to[0].value = value; 
			to.value = value;
			
			o.JsHttpRequest.request('_'+to.name+'_update', to.form);
			o.setFocus(to.name);
		}
	}
	close();
}

/*
	Normalize date format using previous input value to guess missing date elements.
	Helps fast date input field change with only single or double numbers (for day or day-and-month fragments)
*/
function fix_date(date, last)
{
	var dat = last.split(user.datesep);
	var cur = date.split(user.datesep);
	var day, month, year;



	if (date == "" || date == last) 
		return date;
	if (user.datefmt == 0 || user.datefmt == 3) 
	{
		day = dat[1]; month = dat[0]; year = dat[2];
	} else if (user.datefmt == 1 || user.datefmt == 4){
		day = dat[0]; month = dat[1]; year = dat[2];
	} else {
		day = dat[2]; month = dat[1]; year = dat[0];
	}
	if (cur[1] != undefined && cur[1] != "") 
	{
		if (user.datefmt == 0 || user.datefmt == 3 || ((user.datefmt == 2 || user.datefmt == 5) && (cur[2] == undefined || cur[2] == "")))
			day = cur[1];
		else
			month = cur[1];
	}
	if (cur[0] != undefined && cur[0] != "") 
	{
		if (cur[1] == undefined || cur[1] == "")
			day = cur[0];
		else if (user.datefmt == 0 || user.datefmt == 3 || ((user.datefmt == 2 || user.datefmt == 5) && (cur[2] == undefined || cur[2] == "")))
			month = cur[0];
		else if (user.datefmt == 2 || user.datefmt == 5)
			year = cur[0];
		else
			day = cur[0];
	}
	if (cur[2] != undefined && cur[2] != "") 
	{
		if (user.datefmt == 2 || user.datefmt == 5)
			day = cur[2];
		else
			year = cur[2];
	}
	if (user.datefmt<3) {
		if (day<10) day = '0'+parseInt(day, 10);
		if (month<10) month = '0'+parseInt(month, 10);
	}
	if (year<100) year = year<60 ? (2000+parseInt(year,10)) : (1900+parseInt(year,10));

//	console.info(day,month,year)
	if (user.datefmt == 0 || user.datefmt==3)
		return month+user.datesep+day+user.datesep+year;
	if (user.datefmt == 1 || user.datefmt==4)
		return day+user.datesep+month+user.datesep+year;
	return year+user.datesep+month+user.datesep+day;
}

/*
 Behaviour definitions
*/
var inserts = {
	'input': function(e) {
		if(e.onfocus==undefined) {
			e.onfocus = function() {
				save_focus(this);
				if (string_contains(this.className, 'combo') || string_contains(this.className, 'combo3'))
					this.select();
			};
		}
		if (string_contains(e.className, 'combo') || string_contains(e.className, 'combo2') || string_contains(e.className, 'combo3')) {
				_set_combo_input(e);
		}
		else
    		if(e.type == 'text' ) {
   	  			e.onkeydown = function(ev) {
  					ev = ev||window.event;
  					key = ev.keyCode||ev.which;
 	  				if(key == 13) {
						if(e.className == 'searchbox') e.onblur();
						return false;
					}
					return true;
	  			}
			}
	},
	'input.combo2,input[aspect="fallback"]':
	function(e) {
  	    
	    e.style.display = 'none';
	},
	'div.js_only':
	function(e) {
  	    
	    e.style.display = 'block';
	},
	'button': function(e) {
		e.onclick = function(){
			if (validate(e)) {
				setTimeout(function() {	var asp = e.getAttribute('aspect');
					if (asp && asp.indexOf('download') === -1)
						set_mark((asp && ((asp.indexOf('process') !== -1) || (asp.indexOf('nonajax') !== -1))) ? 'progressbar.gif' : 'ajax-loader.gif');
				}, 100);
				return true;
			}
			return false;
		},
		e.onkeydown = function(ev) {	
			ev = ev||window.event;
 			key = ev.keyCode||ev.which;
	  		if(key == 8 || (key==37 && ev.altKey)) {
				ev.returnValue = false;
  			  	return false;
  			}
		}

	},
//	'.ajaxsubmit,.editbutton,.navibutton': 
	'button.ajaxsubmit,input.ajaxsubmit,input.editbutton,button.editbutton,button.navibutton':
	function(e) {
			e.onclick = function() {
				if (validate(e)) {
					save_focus(e);
					var asp = e.getAttribute('aspect')
					if (asp && (asp.indexOf('process') !== -1))
						JsHttpRequest.request(this, null, 600000); 
					else
						JsHttpRequest.request(this);
				}
				return false;
			}
	},
	'.amount': function(e) {
		if (e.onblur == undefined) {
		  e.setAttribute('_last_val', e.value);
  		  e.onblur = function() {
			var dec = this.getAttribute("dec");
			var val = this.getAttribute('_last_val');
			if (val != get_amount(this.name)) {
				this.setAttribute('_last_val', get_amount(this.name));
				price_format(this.name, get_amount(this.name), dec);
				if (e.className.match(/\bactive\b/))
					JsHttpRequest.request('_'+this.name+'_changed', this.form);
			}
		  };
		}
	},
	'.searchbox': 
		function(e) {
			e.setAttribute('_last_val', e.value);
			e.setAttribute('autocomplete', 'off'); //must be off when calling onblur
  		  	e.onblur = function() {
				var val = this.getAttribute('_last_val');
				if (val != this.value) {
					this.setAttribute('_last_val', this.value);
					JsHttpRequest.request('_'+this.name+'_changed', this.form);
				}
			}
	},
	'.date':
		function(e) {
			e.setAttribute('_last_val', e.value);
			e.setAttribute('autocomplete', 'off');
  		  	e.onblur = function() {
				var val = this.getAttribute('_last_val');
				if (val != this.value) {
					this.value = fix_date(this.value, val);
					this.setAttribute('_last_val', this.value);
					if (e.className.match(/\bactive\b/))
						JsHttpRequest.request('_'+this.name+'_changed', this.form);
				}
			}
	},
	'button[aspect*selector], button[aspect*abort], input[aspect*selector]': function(e) {
		e.onclick = function() {
			passBack(this.getAttribute('rel'));
			return false;
		}
	},
	'button[aspect=popup]': function(e) {
		e.onclick = function() {
			if(_w) _w.close(); 
			 var left = (screen.width - 800)/2;
			 var top = (screen.height - 600)/2;
			  _w = open(document.location+'popup=1',
				  "edit","Scrollbars=0,resizable=0,width=800,height=600, top="+top+",left="+left+",screenX="+left+",screenY="+top);
			  if (_w.opener == null)
				  _w.opener = self;
			
//			  _w.moveTo(50, 50);
			  _w.focus();
			return false;
		}
	},
	'select': function(e) {
		if(e.onfocus==undefined) {
			e.onfocus = function() {
			    save_focus(this);
			};
		}
  		var c = e.className;
		if (string_contains(c, 'combo') || string_contains(c, 'combo2') || string_contains(c, 'combo3'))
			_set_combo_select(e);
		else {
			e.onkeydown = function(ev) {	
				ev = ev||window.event;
 				key = ev.keyCode||ev.which;
	  			if(key == 8 || (key=37 && ev.altKey)) {
					ev.returnValue = false;
	  			  	return false;
  				}
			}
		}
	},
	'a.printlink': 	function(l) {
		l.onclick = function() {
		    save_focus(this);
			JsHttpRequest.request(this, null, 60000);
			return false;
		}
	},
	'a.repopts_link': 	function(l) {
		l.onclick = function() {
		    save_focus(this);
		    var replinks = document.getElementsBySelector('a.repopts_link');
				for(var i in replinks)
					replinks[i].style.fontWeight = replinks[i]==this ? 'bold' : 'normal';
			JsHttpRequest.request(this, null);
			return false;
		}
	},
	'a': function(e) { 
  		e.onkeydown = function(ev) {
			ev = ev||window.event;
			key = ev.keyCode||ev.which;
			if(key==37 || key==38 || key==39 || key==40) {
					move_focus(key, e, document.links);
					ev.returnValue = false;
					return false;
			}
		}
		
		if (e.className == 'shortcut'
		 || e.className == 'menu_option'
		 || e.className == 'menu_tab'
 		 || e.className == 'selected')
			e.onclick = function(ev) {
				if (_validate._processing
				 && _validate._modified
				 && !confirm(_validate._processing)) {
					ev.returnValue = false;
					return false;
				}
				if (_hotkeys.alt)	
					return false;
 				window.location = e.href;
			}
	},
	'ul.ajaxtabs':	function(ul) {
	    var ulist=ul.getElementsByTagName("li");
	    for (var x=0; x<ulist.length; x++){ //loop through each LI e
		var tab=ulist[x].getElementsByTagName("button")[0];
		    var url = tab.form.action
		    tab.onclick=function(){
		    if (!_hotkeys.alt && !this.disabled)
				_expand(this);
			return false;
		    }
		}
//	    }
	},
	'textarea': function(e) {
		if(e.onfocus==undefined) {
			e.onfocus = function() {
			    save_focus(this);
			};
		}
	}
/* TODO
	'a.date_picker':  function(e) {
	    
	    e.href = date_picker(this.getAttribute('rel'));
	    e.style.display = '';
	    e.tabindex = -1; 
	}
*/
};

function stopEv(ev) {
			if(ev.preventDefault) {
				ev.preventDefault();
				ev.stopPropagation();
			} else {
				ev.returnValue = false;
				ev.cancelBubble = true;
				window.keycode = 0;
			}
			return false;
}
/*
	Modified accesskey system. While Alt key is pressed letter keys moves
	focus to next marked link. Alt key release activates focused link.
*/
function setHotKeys() {
	document.onkeydown = function(ev) {
		ev = ev||window.event;
		key = ev.keyCode||ev.which;
		if (key == 18 && !ev.ctrlKey) {	
			_hotkeys.alt = true;
			_hotkeys.focus = -1;
			return stopEv(ev);
		}
		else if (ev.altKey && !ev.ctrlKey && ((key>47 && key<58) || (key>64 && key<91))) {
			key = String.fromCharCode(key);
			var n = _hotkeys.focus;
			var l = document.getElementsBySelector('[accesskey='+key+']');
			var cnt = l.length;
			_hotkeys.list = l;
			for (var i=0; i<cnt; i++) {
				n = (n+1)%cnt;
				
				if (l[n].accessKey==key && (l[n].offsetWidth || l[n].offsetHeight)) {
					_hotkeys.focus = n;
	    
					var tmp = function() {l[_hotkeys.focus].focus();};
					setTimeout(tmp, 0);
					break;
				}
			}
			return stopEv(ev);
		}
		if((ev.ctrlKey && key == 13) || key == 27) {
			_hotkeys.alt = false; 
			_hotkeys.focus = -1;
			ev.cancelBubble = true;
   			if(ev.stopPropagation) ev.stopPropagation();
			
			for(var j=0; j<this.forms.length; j++) {
				var form = this.forms[j];
				for (var i=0; i<form.elements.length; i++){
					var el = form.elements[i];
					var asp = el.getAttribute('aspect');

					if (!string_contains(el.className, 'editbutton') && (asp && asp.indexOf('selector') !== -1) && (key==13 || key==27)) {
						passBack(key==13 ? el.getAttribute('rel') : false);
						ev.returnValue = false;
						return false;
					}
					if (((asp && asp.indexOf('default') !== -1) && key==13)||((asp && asp.indexOf('cancel') !== -1) && key==27)) {
						if (validate(el)) {
							if (asp.indexOf('nonajax') !== -1)
								el.click();
							else
							if (asp.indexOf('process') !== -1)
								JsHttpRequest.request(el, null, 600000);
							else
								JsHttpRequest.request(el);
						}
						ev.returnValue = false;
						return false;
					}
				}
			}
			ev.returnValue = false;
			return false;
		}
		if (editors!=='undefined' && editors[key]) {
			callEditor(key);
			return stopEv(ev); 
		}
		return true;
	};
	document.onkeyup = function(ev) {
		ev = ev||window.event;
		key = ev.keyCode||ev.which;

		if (_hotkeys.alt==true) {
			if (key == 18) {
				_hotkeys.alt = false;
				if (_hotkeys.focus >= 0) {
					var link = _hotkeys.list[_hotkeys.focus];
					if(link.onclick)
						link.onclick();
					else
						if (link.target=='_blank') {
							window.open(link.href,'','toolbar=no,scrollbar=no,resizable=yes,menubar=no,width=900,height=500');
							openWindow(link.href,'_blank');
						} else
							window.location = link.href;
				}
			return stopEv(ev);
			}
		}
		return true;
	}
}

Behaviour.register(inserts);

Behaviour.addLoadEvent(setFocus);
Behaviour.addLoadEvent(setHotKeys);
